### ISSUE TITLE
expand incorrectly expands closed type family 

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#43]
In the code below, `th-desugar` thinks `F Bool` reduces to `Int`, when it actually reduces to `Char`:

``` haskell
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
module THDesugarBug where

import Language.Haskell.TH
import Language.Haskell.TH.Desugar

type family F (a :: k) :: * where
    F (a :: * -> *) = Int
    F (a :: k)      = Char

-- Proof that this should reduce to a Char
c :: F Bool
c = 'c'

$(return [])

main :: IO ()
main = putStrLn $([t| F Bool |] >>= dsType >>= expand >>= stringE . pprint . typeToTH)
```

```
$ /opt/ghc/8.0.1/bin/runghc THDesugarBug.hs 
GHC.Types.Int
```

My understanding of the documentation of `expand` and `expandUnsoundly` is that this kind of thing should only happen with the latter function. That is, `expand` shouldn't expand the above example at all, since it distinguishes cases by kind.

**Comment by goldfirere** [comment]
This is actually GHC's fault. I've filed GHC#12646.

**Comment by RyanGlScott** [comment]
[GHC Trac #12646](https://ghc.haskell.org/trac/ghc/ticket/12646) will be fixed in GHC 8.2. So as long as you are willing to use a recent enough GHC HEAD, the code above will work properly:

```
λ> :set -XPolyKinds -XTemplateHaskell -XTypeFamilies 
λ> import Language.Haskell.TH
λ> import Language.Haskell.TH.Desugar
λ> type family F (a :: k) :: * where F (a :: * -> *) = Int; F (a :: k) = Char
λ> putStrLn $([t| F Bool |] >>= dsType >>= expand >>= stringE . pprint . typeToTH)
Ghci1.F GHC.Types.Bool
```
